javaslang/src/main/java/javaslang/collection/AbstractMultimap.java

```diff
@@ -0,0 +1,455 @@
+        return (M) Collections.scanLeft(this, zero, operation, Queue.<Tuple2<K, V>>empty(), Queue::append, this::createFromEntries);
+    }
+
+    @Override
+    public Iterator<Multimap<K, V>> sliding(long size) {
+        return sliding(size, 1);
+    }
+
+    @Override
+    public Iterator<Multimap<K, V>> sliding(long size, long step) {
+        return iterator().sliding(size, step).map(this::createFromEntries);
+    }
+
+    @Override
+    public Tuple2<Multimap<K, V>, Multimap<K, V>> span(Predicate<? super Tuple2<K, V>> predicate) {
+        Objects.requireNonNull(predicate, "predicate is null");
+        final Tuple2<Iterator<Tuple2<K, V>>, Iterator<Tuple2<K, V>>> t = iterator().span(predicate);
+        return Tuple.of(createFromEntries(t._1), createFromEntries(t._2));
+    }
+
+    @Override
+    public M tail() {
+        if (isEmpty()) {
+            throw new UnsupportedOperationException("tail of empty Multimap");
+        } else {
+            final Tuple2<K, V> head = head();
+            return remove(head._1, head._2);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public M take(long n) {
+        if (size() <= n) {
+            return (M) this;
+        } else {
+            return (M) createFromEntries(iterator().take(n));
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public M takeRight(long n) {
+        if (size() <= n) {
+            return (M) this;
+        } else {
+            return (M) createFromEntries(iterator().takeRight(n));
+        }
+    }
+
+    @Override
+    public M takeUntil(Predicate<? super Tuple2<K, V>> predicate) {
+        Objects.requireNonNull(predicate, "predicate is null");
+        return takeWhile(predicate.negate());
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public M takeWhile(Predicate<? super Tuple2<K, V>> predicate) {
+        Objects.requireNonNull(predicate, "predicate is null");
+        final Multimap<K, V> taken = createFromEntries(iterator().takeWhile(predicate));
+        return taken.length() == length() ? (M) this : (M) taken;
+    }
+
+    @Override
+    public int hashCode() {
+        return back.hashCode();
+    }
+
+    @Override
+    public String toString() {
+        return mkString(stringPrefix() + "(", ", ", ")");
+    }
+
+    @Override
+    public java.util.Map<K, Collection<V>> toJavaMap() {
+        final java.util.Map<K, Collection<V>> javaMap = new java.util.HashMap<>();
+        for (Tuple2<K, V> t : this) {
+            javaMap.computeIfAbsent(t._1, k -> {
+                if(emptyContainer.get() instanceof Set) {
+                    return new java.util.HashSet<>();
+                } else if(emptyContainer.get() instanceof SortedSet) {
+                        return new java.util.TreeSet<>();
+                } else {
+                    return new java.util.ArrayList<>();
+                }
+            }).add(t._2);
+        }
+        return javaMap;
```

Comment:
I would move the value-provider to the outside of the loop. Then we make the type check only once.

```java
final java.util.Map<K, Collection<V>> javaMap = new java.util.HashMap<>();
final Supplier<Collection<V>> javaContainerSupplier;
if (containerType == SEQ) {
    javaContainerSupplier = java.util.ArrayList::new;
} else if (containerType == SET) {
    javaContainerSupplier = java.util.HashSet::new;
} else if (containerType == SORTED_SET) {
    javaContainerSupplier = java.util.TreeSet::new;
} else {
    throw new IllegalStateException("Unknown ContainerType: " + containerType);
}
for (Tuple2<K, V> t : this) {
    javaMap.computeIfAbsent(t._1, k -> javaContainerSupplier.get()).add(t._2);
}
return javaMap;
```

Don't know if the throw is possible there or if the compiler will complain that `javaContainerSupplier` is uninitialized. If so, we could create a private method getJavaContainerSupplier()...